dbms/src/Storages/StorageMySQL.cpp

```diff
@@ -0,0 +1,187 @@
+    const ColumnDefaults & column_defaults_,
+    const Context & context_)
+    : IStorage(materialized_columns_, alias_columns_, column_defaults_),
+    table_name(table_name_), server(server_), port(port_), database_name(database_name_), mysql_table_name(mysql_table_name_), user_name(user_name_), password(password_), columns(columns_), context_global(context_),
+    pool(database_name_, server_, user_name_, password_, port_)
+{
+    column_map.set_empty_key("");
+    for (const auto & it: *columns)
+    {
+        column_map[it.name] = it.type;
+    }
+}
+
+bool isCompatible(const IAST & where)
+{
+    String name = where.getID();
+    if ((name == "Function_and") || (name == "Function_or"))
+    {
+        const ASTFunction * and_expression = typeid_cast<const ASTFunction *>(&where);
+        for (const auto & expr: and_expression->arguments->children)
+        {
+            if (!isCompatible(*expr.get()))
+                return false;
+        }
+    }
+    else if (name == "Function_equals" || (name == "Function_notEquals") || (name == "Function_greater") || (name == "Function_less")
+        || (name == "Function_lessOrEquals") || (name == "Function_greaterOrEquals"))
+    {
+        const ASTFunction * function = typeid_cast<const ASTFunction *>(&where);
+        const auto & children = function->arguments->children;
+        if ((children.size() != 2) || !isCompatible(*children[0].get()) || !isCompatible(*children[1].get()))
+            return false;
+    }
+    else if (name == "Function_not")
+    {
+        const ASTFunction * function = typeid_cast<const ASTFunction *>(&where);
+        const auto & children = function->arguments->children;
+        if ((children.size() != 1) || !isCompatible(*children[0].get()))
+            return false;
+    }
+    else if ((strncmp(name.c_str(), "Identifier_", 11) == 0) || (strncmp(name.c_str(), "Literal_", 8) == 0))
+        return true;
+    else
+        return false;
+    return true;
+}
+
+void dumpWhere(const IAST & where, std::stringstream & stream)
+{
+    IAST::FormatSettings s(stream, false, true);
+    IAST::FormatState state;
+    IAST::FormatStateStacked frame;
+    where.formatImpl(s, state, frame);
+}
+
+/** Function puts to stream compatible expressions of where statement.
+  * Compatible expressions are logical expressions on logical expressions or comparisons between fields or constants
+  */
+
+void filterWhere(const IAST & where, std::stringstream & stream)
+{
+    String name = where.getID();
+    if (name == "Function_and")
+    {
+        const ASTFunction * and_expression = typeid_cast<const ASTFunction *>(&where);
+        bool first = true;
+        for (const auto & expr: and_expression->arguments->children)
+        {
+            if (isCompatible(*expr.get()))
+            {
+                if (!first)
+                    stream << " AND ";
+                first = false;
+                dumpWhere(*expr.get(), stream);
```

Comment:
If replace `dumpWhere` -> `filterWhere` it would be possible to skip conditions in complex AND like `a AND b AND (c AND d AND <something to skip>)`. Also, it's possible implement the same logic with OR if all parameters isCompatible: `a OR b OR (c AND d AND <something to skip>)`

Reply:
According to ticket, logic is implemented for top level AND